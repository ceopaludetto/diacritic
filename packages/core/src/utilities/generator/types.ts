import type { ResourceGraph } from "../resource";
import type { DiacriticGenerationOptions, Entry, Parser } from "../types";

import { existsSync, mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { EOL } from "node:os";
import { dirname } from "node:path";

import { dset } from "dset";

import { prefixes } from "../loader";

/// keep-sorted
type GenerateTypesOptions = {
	defaultLanguage: string;
	generation: DiacriticGenerationOptions;
	languages: string[];
	namespaces: string[];
	parser: Parser;
	resourceGraph: ResourceGraph;
};

function toUnion(arr: string[]) {
	return arr.map(item => `"${item}"`).join(" | ");
}

function functionFromEntry(entry: Entry) {
	return `(${entry.args.map(item => item.name + ": " + item.type)}) => string`;
}

export function generateTypes({
	defaultLanguage,
	generation,
	languages,
	namespaces,
	parser,
	resourceGraph,
}: GenerateTypesOptions) {
	const entries = resourceGraph.allEntriesForLanguage(defaultLanguage);
	const declarations: string[] = generation?.banner ?? [];

	declarations.push(
		`declare module "${prefixes[1]}registry" {`,
		`\texport const defaultLanguage: ${JSON.stringify(defaultLanguage)};`,
		`\texport const languages: (${toUnion(languages)})[];`,
		`\texport const namespaces: (${toUnion(namespaces)})[];`,
		"",
		`\texport type SupportedLanguages = typeof languages;`,
		`\texport type SupportedNamespaces = typeof namespaces;`,
		`}`,
		`declare module "${prefixes[2]}registry" {`,
		`\texport * from "${prefixes[1]}registry";`,
		`}`,
	);

	declarations.push(
		`declare module "${prefixes[1]}proxy" {`,
		`\ttype Proxy = {`,
	);

	for (const [namespace, files] of Object.entries(entries)) {
		const contents = files.map(file => readFileSync(file, "utf-8")).flatMap(parser.convertFile);
		const structure = {};

		for (const entry of contents) dset(structure, entry.path, functionFromEntry(entry));

		declarations.push(
			`\t\t${namespace}: ${JSON.stringify(structure).replace(/"/g, "")};`,
		);
	}

	declarations.push(
		`\t};`,
		`}`,
		`declare module "${prefixes[2]}proxy" {`,
		`\texport * from "${prefixes[1]}proxy";`,
		`}`,
	);

	const folder = dirname(generation.outFile);
	if (!existsSync(folder)) mkdirSync(folder, { recursive: true });

	writeFileSync(generation.outFile, declarations.join(EOL), "utf-8");
}
