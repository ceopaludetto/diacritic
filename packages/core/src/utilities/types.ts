export type diacriticGenerationOptions = {
	/**
	 * Additional lines on top of the file. Usually, you want to add lint disable comments
	 */
	banner?: string[];

	/**
	 * Where to put the generated typescript declaration file (.d.ts)
	 *
	 * @example "./src/types/translation.d.ts"
	 */
	outFile: string;

};

export type diacriticOptions = {
	/**
	 * Which of the supported languages will be the default language (fallback)
	 *
	 * @example "en"
	 */
	defaultLanguage: string;

	/**
	 * Generation specific options
	 */
	generation: diacriticGenerationOptions;

	/**
	 * Which languages your application will support
	 *
	 * @example ["en", "pt"]
	 */
	languages: string[];

	/**
	 * Parser to be used to convert translation files to typescript functions
	 */
	parser: Parser;

	/**
	 * Glob containing all resources (translation files).
	 * You may also want to specify `language` replacers
	 *
	 * @example ["src/locales/{{language}}.json", "src/locales/{{language}}/*.json"]
	 */
	resources: string[];
};

export type Entry = {
	/**
	 * The function name
	 */
	name: string;

	/**
	 * The function arguments
	 */
	args: { name: string; type: string }[];

	/**
	 * The function return
	 */
	return: string;
};

export type Parser = {
	/**
	 * Name of the parser
	 */
	name: string;

	/**
	 * Characters to be used to interpolate with variables
	 *
	 * @example ["{{", "}}"]
	 */
	interpolation: [left: string, right: string];

	/**
	 * Convert the given raw file to possible typescript functions
	 *
	 * @param content the raw content of the file
	 * @returns an array of entries (parsed functions)
	 */
	convertFile: (content: string) => Entry[];
};
