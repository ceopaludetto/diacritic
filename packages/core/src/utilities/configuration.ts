import type { DiacriticOptions } from "./types";

import { loadConfig } from "unconfig";

export async function loadConfiguration() {
	const { config } = await loadConfig({
		sources: [
			{
				files: "diacritic.config",
				extensions: ["ts", "mts", "cts", "js", "mjs", "cjs", "json", ""],
			},
		],
		merge: false,
	});

	return config as DiacriticOptions;
}
