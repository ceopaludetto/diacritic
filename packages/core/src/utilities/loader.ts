import invariant from "tiny-invariant";

export const prefixes = ["/~translations/", "~translations/", "virtual:translations/",	"virtual/translations/"];

const translationPathRegex = new RegExp(prefixes.map(prefix => "^" + prefix).join("|"));
const languageAndNamespaceRegex = new RegExp(prefixes[0]! + "([\\w]+)\\/?([\\w]*)");

export function isTranslationPath(path: string) {
	return translationPathRegex.test(path);
}

export function normalizeTranslationPath(path: string) {
	return path.replace(translationPathRegex, prefixes[0]!);
}

export function extractLanguageAndNamespace(languages: string[], path: string) {
	const match = path.match(languageAndNamespaceRegex);
	invariant(Array.isArray(match), "Invalid supplied path");

	const [, first, second] = match;
	invariant(first, "At least one sub path should be available");

	if (first === "registry") return { mode: "registry" } as const;

	if (languages.includes(first)) {
		if (second) return { language: first, namespace: second, mode: "languageAndNamespace" } as const;
		return { language: first, mode: "languageOnly" } as const;
	}

	return { namespace: first, mode: "namespaceOnly" } as const;
}

export function resourceWithoutPrefix(path: string) {
	return path.replace(prefixes[0]!, "");
}

export function isResource(file: string, resources: string[]) {
	for (const resource of resources) {
		const regex = new RegExp(resource.replace("{{language}}", "([^/]+)").replaceAll("*", "([^/]+)"));
		if (regex.test(file)) return true;
	}

	return false;
}
