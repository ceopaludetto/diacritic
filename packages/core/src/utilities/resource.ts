import { resolve } from "node:path";

import { globSync } from "glob";

export class ResourceGraph {
	private graph: Record<string, Record<string, string[]>> = {};

	public constructor(languages: string[], resources: string[]) {
		for (const language of languages) {
			for (const resource of resources) {
				const pattern = resolve(resource).replaceAll("{{language}}", language);
				const files = globSync(pattern);

				const regex = new RegExp(pattern.replace("*", "(.*)"));
				for (const file of files) {
					const namespace = file.match(regex)?.[1] ?? "default";

					if (!this.graph[language]) this.graph[language] = {};
					if (!this.graph[language]![namespace]) this.graph[language]![namespace] = [];
					this.graph[language]![namespace]!.push(file);
				}
			}
		}
	}

	public allFiles() {
		return Object.values(this.graph).flatMap(value => Object.values(value)).flat();
	}

	public allEntries() {
		return this.graph;
	}

	public allEntriesForLanguage(language: string) {
		return this.graph[language] ?? {};
	}

	public allEntriesForLanguageAndNamespace(language: string, namespace: string) {
		return this.graph[language]?.[namespace] ?? [];
	}
}
