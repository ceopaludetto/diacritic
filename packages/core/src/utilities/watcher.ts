import type { ResourceGraph } from "./resource";
import type { AsyncSubscription, Event } from "@parcel/watcher";

import { subscribe } from "@parcel/watcher";

type WatchResourcesOptions = {
	resourceGraph: ResourceGraph;
	resources: string[];
	onChange?: (event?: Event) => Promise<void>;
};

export async function watchResources({ resourceGraph, resources, onChange }: WatchResourcesOptions) {
	await onChange?.();

	const subscriptions: AsyncSubscription[] = [];
	for (const folder of resourceGraph.folders) {
		const subscription = await subscribe(folder, async (err, events) => {
			if (err) throw err;

			for (const event of events) {
				if (event.type === "create") resourceGraph.addFile(event.path, resources);
				if (event.type === "delete") resourceGraph.removeFile(event.path, resources);

				await onChange?.(event);
			}
		});

		subscriptions.push(subscription);
	}

	return async () => await Promise.all(subscriptions.map(async subscription => subscription.unsubscribe()));
}
