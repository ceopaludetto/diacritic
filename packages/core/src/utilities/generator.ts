import type { ResourceGraph } from "./resource";
import type { diacriticGenerationOptions, Entry, Parser } from "./types";

import { readFileSync, writeFileSync } from "node:fs";
import { EOL } from "node:os";

import { prefixes } from "./loader";

export function createFunctionFromEntry(interpolation: Parser["interpolation"], entry: Entry) {
	const interpolators = interpolation.map(side => side.replace(/[.*+?^${}()|[\]\\]/g, "\\$&"));
	const regex = new RegExp(`${interpolators[0]}\\s*(\\w+)\\s*\\|\\s*[\\w\\s]*\\s*${interpolators[1]}`, "g");

	return [
		`export function ${entry.name}(${entry.args.map(item => item.name + ": " + item.type)}) {`,
		`\treturn \`${entry.return.replace(regex, (_, content) => `\${${content}}`)}\`;`,
		`}`,
	].join(EOL);
}

export function createExportFromLanguageAndNamespace(language: string, namespace: string) {
	return `export * as ${namespace} from "${prefixes[0]}${language}/${namespace}";`;
}

export function createImportFromLanguageAndNamespace(language: string, namespace: string) {
	return `import * as ${language} from "${prefixes[0]}${language}/${namespace}";`;
}

export function createConditionalFunctionFromEntry(defaultLanguage: string, languages: string[], entry: Entry) {
	const cases = languages
		.flatMap((language, index) => {
			const fallback = language !== defaultLanguage
				? `?? ${defaultLanguage}?.${entry.name}?.(${entry.args.map(item => item.name).join(", ")})`
				: "";

			return [
				index === 0 ? `\t\tcase "${language}":` : `\t\tcase "${language}":`,
				`\t\t\treturn ${language}?.${entry.name}?.(${entry.args.map(item => item.name).join(", ")})${fallback}`,
			];
		})
		.join(EOL);

	return [
		`export function ${entry.name}(${entry.args.map(item => item.name + ": " + item.type)}) {`,
		`\tswitch (r.language()) {`,
		`${cases}`,
		`\t}`,
		`}`,
	].join(EOL);
}

export function createImportExport(imports: string[], exports: string[]) {
	return [...imports, "", ...exports].join(EOL);
}

export function createTypesFromEntry(entry: Entry) {
	return `export function ${entry.name}(${entry.args.map(item => item.name + ": " + item.type).join(", ")}): string;`;
}

export function createRegistry(defaultLanguage: string, languages: string[]) {
	return [
		`export const defaultLanguage: string = ${JSON.stringify(defaultLanguage)};`,
		`export const languages: string[] = ${JSON.stringify(languages)};`,
	].join(EOL);
}

/// keep-sorted
type GenerateTypesOptions = {
	defaultLanguage: string;
	generation: diacriticGenerationOptions;
	parser: Parser;
	resourceGraph: ResourceGraph;
};

export function generateTypes({ defaultLanguage, generation, parser, resourceGraph }: GenerateTypesOptions) {
	const entries = resourceGraph.allEntries();
	const declarations: string[] = generation?.banner ?? [];

	for (const [language, namespaces] of Object.entries(entries)) {
		for (const [namespace, files] of Object.entries(namespaces)) {
			const content = files.map(file => readFileSync(file, "utf-8"));

			const functions = content.flatMap(item => parser.convertFile(item)).map(item => "\t" + createTypesFromEntry(item));

			declarations.push(
				`declare module "${prefixes[1]}${language}/${namespace}" {`,
				`${functions.join(EOL)}`,
				`}`,
				`declare module "${prefixes[2]}${language}/${namespace}" {`,
				`\texport * from "${prefixes[1]}${language}/${namespace}";`,
				`}`,
			);

			if (language === defaultLanguage) {
				declarations.push(
					`declare module "${prefixes[1]}${namespace}" {`,
					`\texport * from "${prefixes[1]}${language}/${namespace}";`,
					`}`,
					`declare module "${prefixes[2]}${namespace}" {`,
					`\texport * from "${prefixes[1]}${namespace}";`,
					`}`,
				);
			}
		}

		const exports = Object.keys(namespaces).map(namespace => `\texport * as ${namespace} from "${prefixes[1]}${language}/${namespace}";`);
		declarations.push(
			`declare module "${prefixes[1]}${language}" {`,
			`${exports.join(EOL)}`,
			`}`,
			`declare module "${prefixes[2]}${language}" {`,
			`\texport * from "${prefixes[1]}${language}";`,
			`}`,
		);
	}

	writeFileSync(generation.outFile, declarations.join(EOL), "utf-8");
}
