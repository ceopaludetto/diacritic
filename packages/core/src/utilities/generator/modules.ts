import type { Entry, Parser } from "../types";

import { EOL } from "node:os";

import { prefixes } from "../loader";

export function createFunctionFromEntry(interpolation: Parser["interpolation"], entry: Entry) {
	const interpolators = interpolation.map(side => side.replace(/[.*+?^${}()|[\]\\]/g, "\\$&"));
	const regex = new RegExp(`${interpolators[0]}\\s*(\\w+)\\s*\\|\\s*[\\w\\s]*\\s*${interpolators[1]}`, "g");

	return [
		`export function ${entry.name}(${entry.args.map(item => item.name + ": " + item.type)}) {`,
		`\treturn \`${entry.return.replace(regex, (_, content) => `\${${content}}`)}\`;`,
		`}`,
	].join(EOL);
}

export function createExportFromLanguageAndNamespace(language: string, namespace: string) {
	return `export * as ${namespace} from "${prefixes[0]}${language}/${namespace}";`;
}

export function createRegistry(defaultLanguage: string, languages: string[], namespaces: string[]) {
	return [
		`export const defaultLanguage = ${JSON.stringify(defaultLanguage)} as const;`,
		`export const languages = ${JSON.stringify(languages)} as const;`,
		`export const namespaces = ${JSON.stringify(namespaces)} as const;`,
		"",
		`const modules = {`,
		languages.flatMap(language => [
			`\t${language}: {`,
			namespaces
				.map(namespace => `\t\t\t${namespace}: async () => import("${prefixes[1]}${language}/${namespace}"),`)
				.join(EOL),
			`\t},`,
		]).join(EOL),
		`}`,
		"",
		`export async function importTranslationModule(language: string, namespace: string) {`,
		`\treturn modules[language][namespace]();`,
		`}`,
	].join(EOL);
}
