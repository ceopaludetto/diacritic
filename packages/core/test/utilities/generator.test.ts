import type { Entry } from "~/utilities/types";

import { describe, expect, it } from "bun:test";

import { createConditionalFunctionFromEntry, createExportFromLanguageAndNamespace, createFunctionFromEntry, createImportExport, createImportFromLanguageAndNamespace } from "~/utilities/generator";

describe("createFunctionFromEntry", () => {
	it("should create a function from an entry correctly", () => {
		const entry: Entry = {
			name: "hello",
			args: [{ name: "name", type: "string" }],
			return: "Hello, {{ name | string }}!",
		};

		expect(createFunctionFromEntry(["{{", "}}"], entry)).toMatchSnapshot();
	});
});

describe("createExportFromLanguageAndNamespace", () => {
	it("should create a export from language and namespace correctly", () => {
		expect(createExportFromLanguageAndNamespace("en", "common")).toMatchSnapshot();
	});
});

describe("createImportFromLanguageAndNamespace", () => {
	it("should create a import from language and namespace correctly", () => {
		expect(createImportFromLanguageAndNamespace("en", "common")).toMatchSnapshot();
	});
});

describe("createConditionalFunctionFromEntry", () => {
	it("should create a conditional function from an entry correctly", () => {
		const entry: Entry = {
			name: "hello",
			args: [{ name: "name", type: "string" }],
			return: "Hello, {{ name | string }}!",
		};

		expect(createConditionalFunctionFromEntry("en", ["en", "pt"], entry)).toMatchSnapshot();
	});
});

describe("createImportExport", () => {
	it("should create a import/export correctly", () => {
		expect(createImportExport(["import * as m from 'module';"], ["export * from 'module';"])).toMatchSnapshot();
	});
});
