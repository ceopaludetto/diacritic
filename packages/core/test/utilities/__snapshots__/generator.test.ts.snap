// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`createFunctionFromEntry should create a function from an entry correctly 1`] = `
"export function hello(name: string) {
	return \`Hello, ${name}!\`
}"
`;

exports[`createExportFromLanguageAndNamespace should create a export from language and namespace correctly 1`] = `"export * as common from "/~translations/en/common";"`;

exports[`createImportFromLanguageAndNamespace should create a import from language and namespace correctly 1`] = `"import * as en from "/~translations/en/common";"`;

exports[`createConditionalFunctionFromEntry should create a conditional function from an entry correctly 1`] = `
"export function hello(name: string) {
	switch (language()) {
		case "en":
			return en.hello(name)
		case "pt":
			return pt.hello(name)
	}
}"
`;

exports[`createImportExport should create a import/export correctly 1`] = `
"import * as m from 'module';

export * from 'module';"
`;
